design/dedicated-osd-pod.md

```diff
@@ -0,0 +1,147 @@
+# Run OSD in its own Pod
+
+## TL;DR
+
+A one-OSD-per-Pod placement should be implemented to improve reliability and resource efficiency for Ceph OSD daemon.
+
+## Background
+
+Currently in Rook 0.7, Rook Operator starts a ReplicaSet to run [`rook osd`](https://github.com/rook/rook/blob/master/cmd/rook/osd.go) command (hereafter referred to as `OSD Provisioner`)  on each storage node. The ReplicaSet has just one replica. `OSD Provisioner` scans and prepares devices, creates OSD IDs and data directories or devices, generates Ceph configuration. At last, `OSD Provisioner` starts all Ceph OSD, i.e. `ceph-osd`, daemons in foreground and tracks `ceph-osd` processes.
+
+As observed, all Ceph OSDs are running in the same Pod.
+
+## Limitations
+
+The limitations of current design are:
+
+- Reliability issue. One Pod for all OSDs doesn't have the highest reliability nor efficiency. If the Pod is deleted, accidentally or during maintenence, all OSDs are down till the ReplicaSet restart. 
+- Efficiency issue. Resource limits cannot be set effectively on the OSDs since the number of osds per in the pod could vary from node to node. The operator cannot make decisions about the topology because it doesn't know in advance what devices are available on the nodes.
+- Tight Ceph coupling. The monolithic device discovery and provisioning code cannot be reused for other backends.
+- Process management issue. Rook's process management is very simple. Using Kubernetes pod management is much more reliable.
+
+
+A more comprehensive discussions can be found at [this issue](https://github.com/rook/rook/issues/1341).
+
+## Proposal   
+
+We propose the following change to address the limitations.
+
+### Terms
+
+- Device Discovery. A Pod that is given a device or directory path filter and use the filter to discover devices on the host. The Pod updates the device orchestration Configmap with the device paths and mark the orchestration status as "DeviceDiscovered". The Pod is created by Rook Operator and runs in a Kubernetes deployment on the storage node. This process is storage backend agnostic. 
+
+- Device Provisioner. A Pod that is given device or directory paths upon start and make backend specific storage types. For instance, the provisioner prepares OSDs for Ceph backend. It is a Kubernetes batch job and exits after the devices are prepared.
+
+### Create new OSDs
+| Sequence |Rook Operator | Device Discovery  | Device Provisioner   | Ceph OSD Deployment | 
+|---|---|---|---|---|
+| 1  | Read devices from cluster CRD and create an OSD Discovery deployment on each storage node  |   |   |
+| 2  | Watch raw device Configmap  | Discover devices that fit those defined in device filters and update raw device Configmap with the device paths and filtering results| | |
```

Comment:
Even if device discovery runs in its own deployment (instead of part of the agent), i still see that it should run before the operator. It should start when the agent starts and is long-running to continue discovering changes to devices.

Reply:
I would think we should start the discovery pod similarly to the rook agent pod, with a daemonset that will use built-in node affinity and similar properties to deploy. With #1471, the agent is being factored out to be started with yaml instead of programmatically by the operator. The discovery daemonset would follow the same pattern.